### PR TITLE
Added leak/water detection sensor with ESP32 deep sleep/sensor wake

### PR DESCRIPTION
**Pull request by diepeterpan**
## Description:
Added leak/water detection sensor with ESP32 deep sleep / sensor wake

Developed and tested using an HM-RD sensor on a Wemos D1 R32 (ESP32)

image

image

## Checklist:
  - [x] The pull request is done against the latest development branch
  - [x] Only one feature/fix was added per PR and the code change compiles without warnings
  - [x] I accept the [DCO](https://github.com/1technophile/OpenMQTTGateway/blob/development/docs/participate/development.md#developer-certificate-of-origin).

### CODE REVIEW

File: main/main.ino [hunk 1]
```diff
@@ -779,11 +782,33 @@ void setup() {
...
+
 #  ifdef ESP8266_DEEP_SLEEP_IN_US
   Log.notice(F("Setting wake pin for deep sleep." CR));
   pinMode(ESP8266_DEEP_SLEEP_WAKE_PIN, WAKEUP_PULLUP);
 #  endif
 
+#  ifdef ESP32_DEEP_SLEEP_IN_US
+  Log.notice(F("Setting duration for deep sleep." CR));
+  if (esp_sleep_enable_timer_wakeup(ESP32_DEEP_SLEEP_IN_US) != ESP_OK) {
+    Log.notice(F("Failed to set deep sleep duration." CR));
```
Comment (1technophile) [inline main/main.ino]: We could have a `Log.error` here instead of a `Log.notice`

File: main/main.ino [hunk 2]
```diff
@@ -779,11 +782,33 @@ void setup() {
...
+  }
+#  endif
+
+#  ifdef ESP32_EXT0_WAKE_PIN
+  Log.notice(F("Setting EXT0 Wakeup for deep sleep." CR));
+  if (esp_sleep_enable_ext0_wakeup(ESP32_EXT0_WAKE_PIN, ESP32_EXT0_WAKE_PIN_STATE) != ESP_OK) {
+    Log.notice(F("Failed to set deep sleep EXT0 Wakeup." CR));
```
Comment (1technophile) [inline main/main.ino]: Same here

File: main/main.ino [hunk 3]
```diff
@@ -1794,13 +1826,25 @@ void loop() {
...
 /**
- * Deep-sleep for the ESP8266 - e.g. ESP8266_DEEP_SLEEP_IN_US 30000000 for 30 seconds.
+ * Deep-sleep for the ESP8266 & ESP32 - e.g. ESP8266_DEEP_SLEEP_IN_US or ESP32_DEEP_SLEEP_IN_US 30000000 for 30 seconds / wake by ESP32_EXT0_WAKE_PIN.
  * Everything is off and (almost) all execution state is lost.
  */
-#ifdef ESP8266_DEEP_SLEEP_IN_US
+#if defined(ESP8266_DEEP_SLEEP_IN_US) || defined(ESP32_DEEP_SLEEP_IN_US) || defined(ESP32_EXT0_WAKE_PIN)
   if (ready_to_sleep) {
+    delay(250); //Give some time for last MQTT messages to be sent
+#  ifdef ESP8266_DEEP_SLEEP_IN_US
```
Comment (1technophile) [inline main/main.ino]: We could have only one macro that define the deep sleep time to reduce the code:
`DEEP_SLEEP_IN_US`
And use the ESP32 or ESP8266 function depending on the board

File: main/main.ino (hunk 1 above)
Comment (diepeterpan) [reply main/main.ino]: Thank you for spotting it.

File: main/main.ino (hunk 2 above)
Comment (diepeterpan) [reply main/main.ino]: Thank you for spotting it.

File: main/main.ino (hunk 3 above)
Comment (diepeterpan) [reply main/main.ino]: Refactored

**Comment by 1technophile** [comment]
Thanks

**Pull request merged**